File: apps/windows/TeslaSync.App/modals-dialogs/FlagEditDrawer/FlagEditDrawer.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.ModalsDialogs;

/// <summary>
/// One feature-flag row in the shape the editor binds to — the native mirror of the web
/// <c>FeatureFlagEntry</c> (web/src/types/admin-diagnostics.ts: <c>{ key, value }</c>). The backend stores the
/// value as arbitrary JSONB, so the web surfaces it as <c>unknown</c>; the native carries it as a parsed
/// <see cref="System.Text.Json.JsonElement"/> so any JSON value (object / array / scalar / null) round-trips
/// faithfully. The drawer receives this as its <c>initial</c> input (null = "create new" mode).
/// </summary>
public sealed record FeatureFlagEntry(string Key, JsonElement Value)
{
    /// <summary>
    /// Build an entry from a key and a raw JSON string, parsing + cloning the value so it is independent of the
    /// transient parse document. Throws <see cref="JsonException"/> when the value is not valid JSON.
    /// </summary>
    public static FeatureFlagEntry FromJson(string key, string valueJson)
    {
        ArgumentNullException.ThrowIfNull(key);
        ArgumentNullException.ThrowIfNull(valueJson);
        using var document = JsonDocument.Parse(valueJson);
        return new FeatureFlagEntry(key, document.RootElement.Clone());
    }
}

/// <summary>
/// The payload the drawer emits when the operator saves — the native mirror of the object the web passes to
/// <c>onSave({ key, value, reason })</c>. <see cref="Key"/> and <see cref="Reason"/> are already trimmed; the
/// backend audit row rejects an empty reason, which is why the save gate requires it. <see cref="ValueJson"/>
/// is the canonical compact serialization of the parsed value (the body the set-flag mutation posts).
/// </summary>
public sealed record FlagEditSaveRequest(string Key, JsonElement Value, string Reason)
{
    /// <summary>The canonical compact JSON serialization of <see cref="Value"/> (the wire body).</summary>
    public string ValueJson => JsonSerializer.Serialize(Value);
}

/// <summary>
/// The result of parsing the free-form value textarea — the native mirror of the web <c>parsed</c> memo
/// (<c>{ ok, value?, error? }</c>). On success <see cref="Ok"/> is true and <see cref="Value"/> carries the
/// parsed JSON; on failure <see cref="Ok"/> is false and <see cref="Error"/> carries the localized helper text
/// (the empty-value or invalid-JSON message) that disables the save button and renders under the field.
/// </summary>
public sealed class FlagValueParse
{
    private FlagValueParse(bool ok, JsonElement value, string? error)
    {
        Ok = ok;
        Value = value;
        Error = error;
    }

    /// <summary>True when the textarea held a parseable JSON value.</summary>
    public bool Ok { get; }

    /// <summary>The parsed JSON value when <see cref="Ok"/> is true; <c>default</c> otherwise.</summary>
    public JsonElement Value { get; }

    /// <summary>The localized parse-error helper text when <see cref="Ok"/> is false; <c>null</c> otherwise.</summary>
    public string? Error { get; }

    /// <summary>A successful parse carrying the parsed JSON value.</summary>
    public static FlagValueParse Valid(JsonElement value) => new(true, value, null);

    /// <summary>A failed parse carrying the localized helper text.</summary>
    public static FlagValueParse Invalid(string error) => new(false, default, error);
}

/// <summary>
/// Canonical metadata + localized copy for the feature-flag edit / create drawer — the native mirror of the web
/// component's identity and every <c>t('admin.flags.*')</c> / <c>t('common.cancel')</c> call site
/// (web/src/features/admin/components/feature-flags/FlagEditDrawer.tsx). Carries the diagnostics surface slug,
/// the (non-localized) JSON example hint the web hardcodes, and a label helper per i18n key. Every key is
/// the catalog key (<c>translation.*</c> in <c>Strings/{lang}/Resources.resw</c>, P1/S10) and every fallback is
/// the web default; the two interpolated keys (edit title, invalid-JSON) resolve to a <c>{0}</c> format string
/// (the resw form of the web <c>{{key}}</c> / <c>{{msg}}</c> token) and are spliced with
/// <see cref="string.Format(IFormatProvider, string, object?)"/>.
/// </summary>
public static class FlagEditDrawerRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "FlagEditDrawer";

    /// <summary>
    /// The value textarea's example JSON hint. The web hardcodes this JSON snippet inline (it is not a
    /// translated string), so the native mirrors it verbatim rather than routing it through the i18n facade.
    /// </summary>
    public const string ValuePrompt = "{\n  \"enabled\": true\n}";

    /// <summary>Drawer title in "create new" mode (web <c>admin.flags.drawer.createTitle</c>).</summary>
    public static string CreateTitle(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.drawer.createTitle", "Create flag");

    /// <summary>Drawer title in "edit existing" mode, interpolated with the flag key (web <c>admin.flags.drawer.editTitle</c>).</summary>
    public static string EditTitle(ILocalizer localizer, string key) =>
        string.Format(
            CultureInfo.CurrentCulture,
            Require(localizer).GetString("translation.admin.flags.drawer.editTitle", "Edit flag \"{0}\""),
            key ?? string.Empty);

    /// <summary>Save button label (web <c>admin.flags.drawer.save</c>).</summary>
    public static string SaveLabel(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.drawer.save", "Save flag");

    /// <summary>Cancel button label (web <c>common.cancel</c>).</summary>
    public static string CancelLabel(ILocalizer localizer) =>
        Require(localizer).GetString("translation.common.cancel", "Cancel");

    /// <summary>Flag-key field label (web <c>admin.flags.editor.keyLabel</c>).</summary>
    public static string KeyLabel(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.editor.keyLabel", "Flag key");

    /// <summary>Flag-key field input hint (shown when empty).</summary>
    public static string KeyPrompt(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.editor.keyPlaceholder", "feature.dlq.replay_enabled"); // parity:allow web i18n key kept verbatim for catalog parity

    /// <summary>Immutable-key note shown in edit mode (web <c>admin.flags.editor.keyImmutable</c>).</summary>
    public static string KeyImmutableNote(ILocalizer localizer) =>
        Require(localizer).GetString(
            "translation.admin.flags.editor.keyImmutable",
            "Flag keys are immutable once created. Delete + re-create to rename.");

    /// <summary>Value field label (web <c>admin.flags.editor.valueLabel</c>).</summary>
    public static string ValueLabel(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.editor.valueLabel", "Value (JSON)");

    /// <summary>Reason field label (web <c>admin.flags.editor.reasonLabel</c>).</summary>
    public static string ReasonLabel(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.editor.reasonLabel", "Reason");

    /// <summary>Reason field input hint (shown when empty).</summary>
    public static string ReasonPrompt(ILocalizer localizer) =>
        Require(localizer).GetString(
            "translation.admin.flags.editor.reasonPlaceholder", // parity:allow web i18n key kept verbatim for catalog parity
            "Why this change? (logged in audit)");

    /// <summary>Empty-value helper text (web <c>admin.flags.editor.valueEmpty</c>).</summary>
    public static string ValueEmptyError(ILocalizer localizer) =>
        Require(localizer).GetString("translation.admin.flags.editor.valueEmpty", "Value is required.");

    /// <summary>Invalid-JSON helper text, interpolated with the parser message (web <c>admin.flags.editor.valueInvalid</c>).</summary>
    public static string ValueInvalidError(ILocalizer localizer, string message) =>
        string.Format(
            CultureInfo.CurrentCulture,
            Require(localizer).GetString("translation.admin.flags.editor.valueInvalid", "Invalid JSON: {0}"),
            message ?? string.Empty);

    private static ILocalizer Require(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer;
    }
}

/// <summary>
/// UI-free projection logic for the feature-flag editor — the native mirror of the web component's
/// <c>defaultValueJson</c> seed, the <c>parsed</c> JSON memo, and the <c>keyValid</c> / <c>reasonValid</c> /
/// <c>canSave</c> gate. Pure and side-effect free so every branch is asserted headlessly.
/// </summary>
public static class FlagEditDrawerProjection
{
    private static readonly JsonSerializerOptions IndentedOptions = new() { WriteIndented = true };

    /// <summary>
    /// The seed for the value textarea — the web <c>defaultValueJson(initial)</c>: an empty string in "create"
    /// mode, otherwise the entry's value pretty-printed with two-space indentation (web
    /// <c>JSON.stringify(initial.value, null, 2)</c>).
    /// </summary>
    public static string DefaultValueJson(FeatureFlagEntry? initial) =>
        initial is null ? string.Empty : JsonSerializer.Serialize(initial.Value, IndentedOptions);

    /// <summary>
    /// Parse the value textarea — the web <c>parsed</c> memo. An empty / whitespace value is the "value
    /// required" branch; otherwise the text is parsed as JSON, succeeding for any JSON value and surfacing the
    /// parser message on failure. <paramref name="localizer"/> supplies the localized helper text.
    /// </summary>
    public static FlagValueParse ParseValue(string? raw, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        string text = raw ?? string.Empty;
        if (text.Trim().Length == 0)
        {
            return FlagValueParse.Invalid(FlagEditDrawerRegistration.ValueEmptyError(localizer));
        }

        try
        {
            using var document = JsonDocument.Parse(text);
            return FlagValueParse.Valid(document.RootElement.Clone());
        }
        catch (JsonException ex)
        {
            return FlagValueParse.Invalid(FlagEditDrawerRegistration.ValueInvalidError(localizer, ex.Message));
        }
    }

    /// <summary>True once the trimmed key is non-empty (web <c>keyInput.trim().length &gt; 0</c>).</summary>
    public static bool IsKeyValid(string? key) => (key ?? string.Empty).Trim().Length > 0;

    /// <summary>True once the trimmed reason is non-empty (web <c>reason.trim().length &gt; 0</c>).</summary>
    public static bool IsReasonValid(string? reason) => (reason ?? string.Empty).Trim().Length > 0;

    /// <summary>
    /// The save gate — the web <c>canSave = parsed.ok &amp;&amp; keyValid &amp;&amp; reasonValid &amp;&amp; !saving</c>.
    /// </summary>
    public static bool CanSave(bool valueOk, string? key, string? reason, bool saving) =>
        valueOk && IsKeyValid(key) && IsReasonValid(reason) && !saving;
}

/// <summary>
/// PII-safe diagnostics for the <c>FlagEditDrawer</c> surface (P1/S11 diagnostics contract). The drawer edits
/// privileged feature-flag values and an audited reason, so the collector records only the operational
/// <c>view.opened</c> event with the surface slug — never a flag key, value, or reason. Thread-safe.
/// </summary>
public sealed class FlagEditDrawerDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public FlagEditDrawerDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the drawer has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the drawer was opened, emitting <c>view.opened slug=FlagEditDrawer</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(string.Create(CultureInfo.InvariantCulture, $"view.opened slug={FlagEditDrawerRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/modals-dialogs/FlagEditDrawer/FlagEditDrawerModel.kt
// Pure, framework-free model + projection for the FlagEditDrawer modal/dialog surface — the native
// analogue of everything the web component derives before it returns JSX
// (web/src/features/admin/components/feature-flags/FlagEditDrawer.tsx). No Compose, no Android, no
// HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest gate, so
// the composable stays a thin render layer over these pure functions.
//
// The web component is a single controlled create/edit form for one typed feature flag. It binds NO
// data hook — its only dependency is `useTranslation` — so, exactly like the sibling ConfirmDialog /
// GeofenceDrawer surfaces, the cache-then-network lifecycle (loading / empty / error / stale /
// offline) lives on the OWNING page (the FeatureFlags page that holds `useSetFlag` + the `saving`
// flag and passes `initial` / `onSave`), not here; modelling those phases would invent behaviour the
// web spec does not have (drift). The branches the web source actually defines are the complete state
// set this surface renders, and each is projected here:
//   1. create vs. edit mode (web `editing = initial !== null`) — selects the title, the key field's
//      editability, and the immutable-key note,
//   2. the seed JSON for the value editor (web `defaultValueJson` = `JSON.stringify(value, null, 2)`),
//   3. the live parse of the value editor (web `parsed` memo): empty → "value required", unparseable
//      → "invalid JSON" + the parser message, otherwise the decoded JSON value,
//   4. the composite save-enabled rule (web `canSave = parsed.ok && keyValid && reasonValid &&
//      !saving`) and the trimmed `{ key, value, reason }` payload handed back through `onSave`.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/modals-dialogs/FlagEditDrawer — the P3 prompt's allowed-files path) cannot form a
// valid Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the
// package intentionally diverges from the path — exactly as the sibling ConfirmDialog / GeofenceDrawer
// surfaces do. `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.modalsdialogs.flageditdrawer

import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object FlagEditDrawerRegistration {
    /** Stable surface id. */
    const val ID: String = "flag-edit-drawer"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "FlagEditDrawer"
}

/**
 * The flag the drawer edits — the native mirror of the web `initial: FeatureFlagEntry | null` prop. A
 * `null` target is the web "create new" mode; a non-null target is "edit existing". [value] is
 * arbitrary JSON (the web `FeatureFlagValue = unknown`, stored as JSONB in Postgres), carried as a
 * nullable [JsonElement] where a Kotlin `null` models the web `undefined` (absent — seeds an empty
 * editor) and a JSON `null` models a literal `null` value (seeds the text `null`).
 *
 * @property key the flag key (web `initial.key`); shown read-only in edit mode.
 * @property value the flag's current JSON value (web `initial.value`), or `null` when absent.
 */
data class FlagEditTarget(
    val key: String,
    val value: JsonElement?,
)

/**
 * The assembled, trimmed write the operator confirms — the native mirror of the object the web hands
 * to `onSave({ key, value, reason })`. The owning page forwards it to the shared `setFlag` mutation
 * (S8); the drawer itself performs no HTTP.
 *
 * @property key the trimmed flag key (web `keyInput.trim()`).
 * @property value the decoded JSON value to store (web `parsed.value`).
 * @property reason the trimmed audit reason (web `reason.trim()`), required by the backend audit row.
 */
data class FlagEditSubmission(
    val key: String,
    val value: JsonElement,
    val reason: String,
)

/**
 * The outcome of parsing the free-form JSON value editor — the native mirror of the web `parsed` memo
 * (`{ ok, value?, error? }`). A pure sum type so the composable maps each case to its localized helper
 * text + the save-enabled rule without re-deriving anything: [Empty] when the editor is blank (web
 * `valueEmpty`), [Invalid] when `JSON.parse` throws (web `valueInvalid` with the parser message), and
 * [Valid] carrying the decoded value when it parses.
 */
sealed interface FlagValueParse {
    /** The editor is blank — the web `valueInput.trim() === ''` branch (renders "Value is required."). */
    data object Empty : FlagValueParse

    /**
     * The editor text is not valid JSON — the web `catch` branch. [message] is the parser's own error
     * text, interpolated into the localized "Invalid JSON: {{msg}}" copy at the Compose boundary.
     */
    data class Invalid(
        val message: String,
    ) : FlagValueParse

    /** The editor parsed cleanly — the web `{ ok: true, value }` branch; [value] is the decoded JSON. */
    data class Valid(
        val value: JsonElement,
    ) : FlagValueParse
}

/**
 * The pure derivations the composable renders over — a 1:1 port of the web component's `editing`
 * flag, `defaultValueJson` helper, `parsed` memo, and `canSave` rule. Stateless and side-effect-free,
 * so the surface is fully covered by the off-device unit gate.
 */
object FlagEditDrawerProjection {
    // Web `JSON.stringify(value, null, 2)`: pretty-printed with a two-space indent.
    private val prettyJson =
        Json {
            prettyPrint = true
            prettyPrintIndent = "  "
        }

    // Strict decoder matching `JSON.parse` (no lenient/relaxed parsing): an unquoted key, a trailing
    // token, or a bare word is rejected, exactly as the browser's parser rejects it.
    private val strictJson = Json

    // The parser-detail text shown when a top-level unquoted bareword is rejected (interpolated into the
    // localized "Invalid JSON: {{msg}}" copy, mirroring the web's raw `e.message` engine detail).
    private const val NON_JSON_VALUE_MESSAGE = "Unexpected token in JSON"

    /** Whether the drawer is editing an existing flag vs. creating one — web `initial !== null`. */
    fun isEditing(target: FlagEditTarget?): Boolean = target != null

    /**
     * The text the value editor opens with — a port of the web `defaultValueJson(initial)`: an absent
     * target (create mode) or an absent value seeds an empty editor; otherwise the value is
     * pretty-printed as `JSON.stringify(value, null, 2)`. Any encode failure falls back to an empty
     * editor, mirroring the web `catch { return '' }`.
     */
    fun defaultValueJson(target: FlagEditTarget?): String {
        val value = target?.value ?: return ""
        return runCatching { prettyJson.encodeToString(JsonElement.serializer(), value) }.getOrDefault("")
    }

    /**
     * Parses the value editor exactly as the web `parsed` memo: a blank editor is [FlagValueParse.Empty]
     * (web `valueInput.trim() === ''`), otherwise the raw text is decoded; a decode failure yields
     * [FlagValueParse.Invalid] carrying the parser message (web `catch`), and a clean decode yields
     * [FlagValueParse.Valid] with the JSON value (web `JSON.parse(valueInput)`).
     *
     * Strictness note (no silent drift): kotlinx's lexer accepts a top-level unquoted bareword (e.g.
     * `not-json`) as a non-string primitive, which the browser `JSON.parse` rejects. To keep the surface
     * byte-for-byte faithful to the web — so an operator can never persist a value the web would have
     * blocked — such barewords are re-rejected here via [isJsonParseCompatible].
     */
    fun parseValue(raw: String): FlagValueParse {
        if (raw.isBlank()) return FlagValueParse.Empty
        return try {
            val element = strictJson.parseToJsonElement(raw)
            if (isJsonParseCompatible(element)) {
                FlagValueParse.Valid(element)
            } else {
                FlagValueParse.Invalid(NON_JSON_VALUE_MESSAGE)
            }
        } catch (e: SerializationException) {
            FlagValueParse.Invalid(e.message ?: e.toString())
        }
    }

    /**
     * Whether [element] is something the browser `JSON.parse` would also have accepted. Objects, arrays,
     * and quoted strings always are; a non-string primitive is only valid when it is a recognised JSON
     * literal (`true` / `false` / `null`) or a number — rejecting the unquoted barewords kotlinx's
     * lenient top-level lexer would otherwise admit.
     */
    private fun isJsonParseCompatible(element: JsonElement): Boolean {
        val primitive = element as? JsonPrimitive ?: return true
        val content = primitive.content
        return primitive.isString ||
            content == "true" ||
            content == "false" ||
            content == "null" ||
            primitive.doubleOrNull != null
    }

    /** Whether the flag key is non-blank — the web `keyInput.trim().length > 0` guard. */
    fun isKeyValid(key: String): Boolean = key.trim().isNotEmpty()

    /** Whether the audit reason is non-blank — the web `reason.trim().length > 0` guard. */
    fun isReasonValid(reason: String): Boolean = reason.trim().isNotEmpty()

    /**
     * Whether the Save action is enabled — the web `canSave = parsed.ok && keyValid && reasonValid &&
     * !saving`. The value must parse, the key and reason must be non-blank, and no save may be in
     * flight; otherwise Save stays disabled exactly as the web button does.
     */
    fun canSubmit(
        parse: FlagValueParse,
        key: String,
        reason: String,
        saving: Boolean,
    ): Boolean = parse is FlagValueParse.Valid && isKeyValid(key) && isReasonValid(reason) && !saving

    /**
     * Assembles the trimmed `{ key, value, reason }` write the operator confirms — the web
     * `onSave({ key: keyInput.trim(), value: parsed.value, reason: reason.trim() })`. The caller passes
     * the already-decoded [value] from a [FlagValueParse.Valid], so this never re-parses.
     */
    fun buildSubmission(
        key: String,
        value: JsonElement,
        reason: String,
    ): FlagEditSubmission = FlagEditSubmission(key = key.trim(), value = value, reason = reason.trim())
}

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [FlagEditDrawerRegistration.SLUG]
 * (P1/S11). Carries only the slug — never the flag key, value, or reason — so a diagnostics line can
 * never leak the server's feature-flag posture or an operator's draft. Kept free of Compose so it is
 * unit-tested with a recording [Logger]; the composable calls it from its first-composition effect.
 */
fun recordFlagEditDrawerOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to FlagEditDrawerRegistration.SLUG))
}
